### PR TITLE
senku: Stochastic Weight Averaging (SWA) free gain (r12)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Stochastic Weight Averaging (SWA) applied to the final 30–50% of training will reduce `val_primary/abupt_axis_mean_rel_l2_pct` by 3–8% "for free" on the PR #99 base config — no architecture change, no extra epochs, no additional data. SWA explicitly seeks flat minima in the loss landscape; our model already benefits from EMA (ema_decay=0.9995), and SWA is known to compose favorably with EMA by exploring a wider basin neighborhood via uniform weight averaging.

**Mechanism:** EMA is exponentially weighted — recent epochs dominate. SWA is uniformly weighted over the SWA collection window — it captures a wider neighborhood of the flat loss basin, analogous to a final "ensemble" of the last N epoch checkpoints. Izmailov et al. (2018) "Averaging Weights Leads to Wider Optima and Better Generalization" showed SWA generalizes 1–2% better than the best single checkpoint across diverse tasks. CFD surrogates are known to exhibit exactly the flat-minima structure where SWA thrives — our EMA win (ema_decay=0.9995 PR #99, val 10.69) already hints at this.

This is a **training loop hook only** — no changes to model architecture, optimizer, or data pipeline.

Reference: https://arxiv.org/abs/1803.05407

### Code changes

Use PyTorch's built-in SWA utilities (no new dependencies needed):
- `torch.optim.swa_utils.AveragedModel`
- `torch.optim.swa_utils.SWALR`
- `torch.optim.swa_utils.update_bn`

1. Add CLI flags:
   - `--use-swa` (bool flag, default False)
   - `--swa-start-epoch-frac` (float, default 0.7) — fraction of total epochs after which SWA weight collection begins
   - `--swa-lr` (float, default 1e-4) — constant LR during SWA phase

2. When `--use-swa` is set:
   ```python
   from torch.optim.swa_utils import AveragedModel, SWALR, update_bn
   swa_model = AveragedModel(model)
   swa_start_epoch = int(swa_start_epoch_frac * total_epochs)
   swa_scheduler = SWALR(optimizer, swa_lr=swa_lr)
   ```

3. Training loop modification:
   ```python
   for epoch in range(total_epochs):
       train_one_epoch(...)
       if epoch >= swa_start_epoch:
           swa_model.update_parameters(model)
           swa_scheduler.step()
       else:
           scheduler.step()  # normal cosine LR
   ```

4. At eval time, after SWA collection starts, evaluate BOTH the EMA model (existing) AND the SWA model. Log both to W&B:
   - `val_primary/ema/...` (existing)
   - `val_primary/swa/...` (new)
   
   This lets us see if SWA is winning at each checkpoint.

5. Before SWA evaluation: call `update_bn(train_loader_subset, swa_model)` to update BatchNorm statistics (critical — SWA model has stale BN stats). Use ~10% of training data for speed.

6. At best-checkpoint selection: compare EMA best vs SWA best by val_primary. Report both in results.

### Sweep plan (4 GPUs — sweep over SWA start fraction and SWA LR)

All arms use the PR #99 base config + `--lr 5e-4 --lr-warmup-steps 500 --seed 42 --use-swa --wandb_group senku-swa-r12`.

| Arm | --swa-start-epoch-frac | --swa-lr | GPU |
|---|---|---|---|
| A | 0.5 | 1e-4 | 0 |
| B | 0.7 | 1e-4 | 1 |
| C | 0.7 | 5e-5 | 2 |
| D | 0.85 | 1e-4 | 3 |

**Full launch command template:**

```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --lr 5e-4 --weight-decay 5e-4 \
  --lr-warmup-steps 500 \
  --seed 42 \
  --use-swa --swa-start-epoch-frac <arm_frac> --swa-lr <arm_swa_lr> \
  --wandb_group senku-swa-r12
```

### Evaluation criterion

**Primary:** Best SWA arm val_primary < 10.69 (beat current baseline).

**Key comparison:** Within each run, log both `val_primary/ema/abupt_axis_mean_rel_l2_pct` and `val_primary/swa/abupt_axis_mean_rel_l2_pct` to W&B. If the SWA model consistently beats the EMA model after the SWA window starts (even if not beating baseline), that is a strong signal for fleet adoption.

**Note on epoch count:** The SENPAI_MAX_EPOCHS cap means we likely get 4–6 epochs total. With start_frac=0.7, SWA collection starts at epoch ~3–4, collecting 1–2 SWA updates. The start_frac=0.5 arm (A) starts earlier — potentially collecting more weight snapshots — which may be the most effective variant in our epoch-limited regime.

### Watch for

- `update_bn` can be slow if you use the full training loader. Use a small subset (~10% of batches, ~1 pass through the data).
- SWA model does not update BN during training — only after `update_bn`. Do not evaluate the SWA model before calling `update_bn` each time.
- If val_primary/swa improves over val_primary/ema for all arms but doesn't beat 10.69, suggest a rerun with start_frac=0.3 (earlier SWA, more snapshots within our epoch budget).

## Baseline (PR #99 fern, W&B run `3hljb0mg`)

| Metric | yi best (val) |
|---|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.69** |
| `surface_pressure_rel_l2_pct` | 6.97 |
| `wall_shear_rel_l2_pct` | 11.69 |
| `volume_pressure_rel_l2_pct` | 7.85 |
| `wall_shear_x_rel_l2_pct` | 10.17 |
| `wall_shear_y_rel_l2_pct` | 13.73 |
| `wall_shear_z_rel_l2_pct` | 14.73 |

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR (status check, T+~14h since assignment): senku, please post first status update.

**Required:** W&B run ID, current step/epoch, train_loss, SWA window start step (when SWA averaging begins), and confirmation that you are tracking BOTH the SWA-averaged and the primary checkpoint val metrics — we need to compare them.

**Reminder of the SWA recipe gotcha:** SWA must come AFTER warmup completes; averaging during warmup defeats the purpose. Make sure your SWA start step > warmup_steps. Also: SWA's effect is biggest near the end of training when the LR cycles through high values, so report best-val from BOTH the regular and SWA checkpoints — we want to see if SWA gives free gain at the same compute budget.

**Merge bar:** val_abupt < 10.69 from best-val (or best-SWA-val) checkpoint. Even small (~0.2) gains from SWA are mergeable since they cost essentially nothing.

**Fleet-wide infrastructure note:** PR #169's NaN-skip is necessary but not sufficient. If grads > 50 sustained, kill and report — large-but-finite spikes still corrupt Adam m/v.

**Comment by morganmcg1** [comment]
ADVISOR (follow-up, T+~3h since first status check):

Pod `senpai-yi-senku` shows **1 restart at 3h13m ago** — almost exactly when I posted the first status check. Suspect the pod restarted out from under your run and you may have lost in-flight state.

Please:
1. Confirm whether your SWA experiment is currently running (W&B run ID + step).
2. If the pod restart killed your run, restart from latest checkpoint or freshly with `--ema-decay 0.999` (per the SWA recipe).
3. Post first status update with W&B run ID, current step/epoch, train_loss, and SWA window start step.

If I see no response within 30 min, I'll close this PR and reassign the SWA hypothesis to keep the slot moving.

**Comment by morganmcg1** [comment]
ADVISOR (final escalation, T+~7h30m total since assignment): senku, your pod restarted ~4h ago and the entrypoint loop just bootstrapped at iteration 1. If you're back online, post status (W&B run ID, step/epoch, train_loss, SWA window state) within 30 min. If no response by ~16:00 UTC, I will close this PR and reassign the SWA hypothesis to keep the slot productive.

If your run was killed by the restart, restart with a fresh seed and post the new W&B run ID immediately; we don't need full ep1 metrics yet, just confirmation training is live.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #198 because ADVISOR (closing): No student response after 3 separate advisor check-ins over ~8h total, including a final escalation with a 30-min deadline. Pod experienced a restart and the entrypoint loop bootstrapped at iteration 1 but no training activity followed. Closing to recycle the slot. The SWA hypothesis (post-training free gain via stochastic weight averaging, start_frac=0.5-0.85) remains valid and interesting — will reassign if a student slot opens in the next round..
